File: packages/i18nify-js/src/modules/geo/__tests__/getCountryTaxDefinition.test.ts
import getCountryTaxDefinition from '../getCountryTaxDefinition';

describe('getCountryTaxDefinition', () => {
  describe('GST countries', () => {
    it('returns GST definition for India (IN)', () => {
      const result = getCountryTaxDefinition('IN');
      expect(result).toEqual({
        tax_name: 'GST',
        full_name: 'Goods and Services Tax',
        standard_rate: 18,
        notes: 'Multiple slabs: 0%, 5%, 12%, 18%, 28%',
      });
    });

    it('returns GST definition for Australia (AU)', () => {
      const result = getCountryTaxDefinition('AU');
      expect(result.tax_name).toBe('GST');
      expect(result.standard_rate).toBe(10);
    });

    it('returns GST definition for Singapore (SG)', () => {
      const result = getCountryTaxDefinition('SG');
      expect(result.tax_name).toBe('GST');
      expect(result.standard_rate).toBe(9);
    });
  });

  describe('VAT countries', () => {
    it('returns VAT definition for Germany (DE)', () => {
      const result = getCountryTaxDefinition('DE');
      expect(result).toEqual({
        tax_name: 'MwSt',
        full_name: 'Mehrwertsteuer',
        standard_rate: 19,
        notes: '',
      });
    });

    it('returns VAT definition for United Kingdom (GB)', () => {
      const result = getCountryTaxDefinition('GB');
      expect(result.tax_name).toBe('VAT');
      expect(result.standard_rate).toBe(20);
    });

    it('returns VAT definition for UAE (AE)', () => {
      const result = getCountryTaxDefinition('AE');
      expect(result.tax_name).toBe('VAT');
      expect(result.standard_rate).toBe(5);
    });
  });

  describe('other tax systems', () => {
    it('returns Sales Tax definition for US', () => {
      const result = getCountryTaxDefinition('US');
      expect(result.tax_name).toBe('Sales Tax');
      expect(result.standard_rate).toBe(0);
      expect(result.notes).toContain('No federal rate');
    });

    it('returns SST definition for Malaysia (MY)', () => {
      const result = getCountryTaxDefinition('MY');
      expect(result.tax_name).toBe('SST');
      expect(result.standard_rate).toBe(10);
    });

    it('returns no-tax definition for Hong Kong (HK)', () => {
      const result = getCountryTaxDefinition('HK');
      expect(result.tax_name).toBe('');
      expect(result.standard_rate).toBe(0);
      expect(result.notes).toContain('No VAT or GST');
    });
  });

  describe('input normalisation', () => {
    it('accepts lowercase country code', () => {
      const result = getCountryTaxDefinition('in');
      expect(result.tax_name).toBe('GST');
    });

    it('accepts mixed-case country code', () => {
      const result = getCountryTaxDefinition('De');
      expect(result.tax_name).toBe('MwSt');
    });
  });

  describe('return shape', () => {
    it('always returns all four fields', () => {
      const result = getCountryTaxDefinition('FR');
      expect(result).toHaveProperty('tax_name');
      expect(result).toHaveProperty('full_name');
      expect(result).toHaveProperty('standard_rate');
      expect(result).toHaveProperty('notes');
    });

    it('standard_rate is a number', () => {
      const result = getCountryTaxDefinition('FI');
      expect(typeof result.standard_rate).toBe('number');
      expect(result.standard_rate).toBe(25.5);
    });
  });

  describe('error handling', () => {
    it('throws for empty string', () => {
      expect(() => getCountryTaxDefinition('')).toThrow(
        "Parameter 'countryCode' is invalid!",
      );
    });

    it('throws for unsupported country code', () => {
      expect(() => getCountryTaxDefinition('XX')).toThrow(
        'Country code "XX" is not supported',
      );
    });

    it('error message for unsupported code lists supported countries', () => {
      expect(() => getCountryTaxDefinition('ZZ')).toThrow(
        'Supported countries:',
      );
    });
  });
});

File: packages/i18nify-go/modules/geo/get_country_tax_definition_test.go
package geo

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestGetCountryTaxDefinition(t *testing.T) {
	tests := []struct {
		name        string
		input       string
		wantTaxName string
		wantRate    float64
		wantErr     bool
		errContains string
	}{
		// GST countries
		{name: "India GST", input: "IN", wantTaxName: "GST", wantRate: 18},
		{name: "Australia GST", input: "AU", wantTaxName: "GST", wantRate: 10},
		{name: "Singapore GST", input: "SG", wantTaxName: "GST", wantRate: 9},
		{name: "New Zealand GST", input: "NZ", wantTaxName: "GST", wantRate: 15},
		// VAT countries
		{name: "Germany MwSt", input: "DE", wantTaxName: "MwSt", wantRate: 19},
		{name: "UK VAT", input: "GB", wantTaxName: "VAT", wantRate: 20},
		{name: "France TVA", input: "FR", wantTaxName: "TVA", wantRate: 20},
		{name: "UAE VAT", input: "AE", wantTaxName: "VAT", wantRate: 5},
		{name: "Hungary highest EU rate", input: "HU", wantTaxName: "ÁFA", wantRate: 27},
		{name: "Finland decimal rate", input: "FI", wantTaxName: "ALV", wantRate: 25.5},
		// Other systems
		{name: "US Sales Tax zero rate", input: "US", wantTaxName: "Sales Tax", wantRate: 0},
		{name: "Malaysia SST", input: "MY", wantTaxName: "SST", wantRate: 10},
		{name: "Japan Consumption Tax", input: "JP", wantTaxName: "CT", wantRate: 10},
		{name: "Brazil IBS/CBS", input: "BR", wantTaxName: "IBS/CBS", wantRate: 26.5},
		// No indirect tax
		{name: "Hong Kong no tax", input: "HK", wantTaxName: "", wantRate: 0},
		{name: "Kuwait no tax", input: "KW", wantTaxName: "", wantRate: 0},
		// Input normalisation
		{name: "lowercase accepted", input: "in", wantTaxName: "GST", wantRate: 18},
		{name: "padded whitespace trimmed", input: "  DE  ", wantTaxName: "MwSt", wantRate: 19},
		// Errors
		{name: "empty string errors", input: "", wantErr: true, errContains: "non-empty"},
		{name: "whitespace only errors", input: "   ", wantErr: true, errContains: "non-empty"},
		{name: "unsupported code errors", input: "XX", wantErr: true, errContains: "not supported"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := GetCountryTaxDefinition(tt.input)
			if tt.wantErr {
				assert.Error(t, err)
				if tt.errContains != "" {
					assert.True(t, strings.Contains(err.Error(), tt.errContains),
						"expected error to contain %q, got %q", tt.errContains, err.Error())
				}
				return
			}
			assert.NoError(t, err)
			assert.Equal(t, tt.wantTaxName, result.TaxName)
			assert.Equal(t, tt.wantRate, result.StandardRate)
		})
	}
}

func TestGetCountryTaxDefinition_FullFields(t *testing.T) {
	result, err := GetCountryTaxDefinition("IN")
	assert.NoError(t, err)
	assert.Equal(t, "GST", result.TaxName)
	assert.Equal(t, "Goods and Services Tax", result.FullName)
	assert.Equal(t, float64(18), result.StandardRate)
	assert.Contains(t, result.Notes, "Multiple slabs")
}

func TestGetCountryTaxDefinition_NotesPresent(t *testing.T) {
	// US has notes about no federal rate
	result, err := GetCountryTaxDefinition("US")
	assert.NoError(t, err)
	assert.Contains(t, result.Notes, "No federal rate")

	// CA has notes about HST
	result, err = GetCountryTaxDefinition("CA")
	assert.NoError(t, err)
	assert.Contains(t, result.Notes, "HST")
}

func TestGetCountryTaxDefinition_NoTaxCountries(t *testing.T) {
	result, err := GetCountryTaxDefinition("HK")
	assert.NoError(t, err)
	assert.Equal(t, "", result.TaxName)
	assert.Equal(t, float64(0), result.StandardRate)
	assert.Contains(t, result.Notes, "No VAT or GST")
}
